internal/datacoord/clustering_compaction_manager.go

```diff
@@ -0,0 +1,809 @@
+		// task must be cleaned properly if task fail or timeout
+	}
+	return nil
+}
+
+func (t *ClusteringCompactionManager) processInitTask(task *datapb.CompactionTask) error {
+	log := log.With(zap.Int64("triggerID", task.TriggerID), zap.Int64("collectionID", task.GetCollectionID()), zap.Int64("planID", task.GetPlanID()))
+	var operators []UpdateOperator
+	for _, segID := range task.InputSegments {
+		operators = append(operators, UpdateSegmentLevelOperator(segID, datapb.SegmentLevel_L2))
+	}
+	err := t.meta.UpdateSegmentsInfo(operators...)
+	if err != nil {
+		log.Warn("fail to set segment level to L2", zap.Error(err))
+		return err
+	}
+
+	if typeutil.IsVectorType(task.GetClusteringKeyField().DataType) {
+		err := t.submitToAnalyze(task)
+		if err != nil {
+			log.Warn("fail to submit analyze task", zap.Error(err))
+			return merr.WrapErrClusteringCompactionSubmitTaskFail("analyze", err)
+		}
+	} else {
+		err := t.submitToCompact(task)
+		if err != nil {
+			log.Warn("fail to submit compaction task", zap.Error(err))
+			return merr.WrapErrClusteringCompactionSubmitTaskFail("compact", err)
+		}
+	}
+	return nil
+}
+
+func (t *ClusteringCompactionManager) processExecutingTask(task *datapb.CompactionTask) error {
+	compactionTask := t.compactionHandler.getCompaction(task.GetPlanID())
+	if compactionTask == nil {
+		// if one compaction task is lost, mark it as failed, and the clustering compaction will be marked failed as well
+		log.Warn("compaction task lost", zap.Int64("planID", task.GetPlanID()))
+		// trigger retry
+		oldPlanID := task.GetPlanID()
+		// todo: whether needs allocate a new planID
+		task.State = datapb.CompactionTaskState_compact_pipelining
+		return merr.WrapErrClusteringCompactionCompactionTaskLost(oldPlanID)
+	}
+	log.Info("compaction task state", zap.Int64("planID", compactionTask.plan.PlanID), zap.Int32("state", int32(compactionTask.state)))
+	task.State = compactionTaskStateV2(compactionTask.state)
+	if task.State == datapb.CompactionTaskState_compact_completed {
+		return t.processCompactedTask(task)
+	}
+	return nil
+}
+
+func (t *ClusteringCompactionManager) processCompactedTask(task *datapb.CompactionTask) error {
+	if len(task.ResultSegments) == 0 {
+		compactionTask := t.compactionHandler.getCompaction(task.GetPlanID())
+		if compactionTask == nil {
+			// if one compaction task is lost, mark it as failed, and the clustering compaction will be marked failed as well
+			log.Warn("compaction task lost", zap.Int64("planID", task.GetPlanID()))
+			// trigger retry
+			oldPlanID := task.GetPlanID()
+			// todo: whether needs allocate a new planID
+			task.State = datapb.CompactionTaskState_compact_pipelining
+			return merr.WrapErrClusteringCompactionCompactionTaskLost(oldPlanID)
+		}
+		segmentIDs := make([]int64, 0)
+		for _, seg := range compactionTask.result.Segments {
+			segmentIDs = append(segmentIDs, seg.GetSegmentID())
+		}
+		task.ResultSegments = segmentIDs
+	}
+
+	return t.processIndexingTask(task)
```

Comment:
so I think maybe we can just remove Compacted state?

Reply:
This is to be compatible with old compaction state. Actually, yes, it compacted and indexing are the same actually. No harm to keep it, right?